Work in progress - Implementing the router

aespasa opened the pull request:
close #5

aespasa committed:
Work in progress

TBoileau commented on config/routes.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "get": {
+    "/": {
+      "name": "homepage"
+    },
+    "posts/:id": {
+      "name": "get_post",
+      "requirements": {
+        "id": "[0-9]+"
+      }
+    }
+  },
+  "post": {
+    "/posts": {
+      "name": "publish_post"
+    }
+  },
+  "patch": {
+    "/posts/:id": {
+      "name": "update_post",
+      "requirements": {
+        "id": "[0-9]+"
+      }
+    }
+  },
+  "delete": {
+    "/posts/:id": {
+      "name": "delete_post",
+      "requirements": {
+        "id": "[0-9]+"
+      }
+    }
+  }
+}
```
Remplzcer ce fichier par un fichier routes.php 
```php
<?php

return [
   new Route(.....),
   new Route(.....)
];
```

TBoileau commented on src/Router/UrlGenerator.php:
```diff
@@ -0,0 +1,10 @@
+
+namespace Blog\Router;
+
+final class UrlGenerator
```
Transformer en UrlGeneratorInterface (ne pas hésiter à regarder la classe Router de Symfony), ça permet de respecter la ségrégation des interfaces

TBoileau commented on src/Router/Router.php:
```diff
@@ -0,0 +1,56 @@
+{
+    private const NO_ROUTE = 404;
+
+    public function __construct(
+        private ArrayIterator $routes,
+        private UrlGenerator $urlGenerator
+    ) {
```
Pas besoin d'injecter d'urlGenerator

TBoileau commented on src/Router/Router.php:
```diff
@@ -0,0 +1,56 @@
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+
+final class Router
```
Router doit implémenter 2 interfaces, UrlGeneratorInterface et RouterInterface

aespasa committed:
Implementing router v1.0

TBoileau commented on src/Router/Route.php:
```diff
@@ -4,56 +4,111 @@
+     * @param string $path
+     * @param array|callable $callable
+     */
+    public function __construct(string $name, string $path, callable|array $callable)
```
Pas besoin de callable, dans notre cas, ce sera toujours controller+action

TBoileau commented on src/Router/Route.php:
```diff
@@ -4,56 +4,111 @@
+
+        if (count($parameters) > 0) {
+            $parameters = array_combine($parameters, $matches);
+
+            if (is_array($this->callable)) {
+                $reflFunc = (new ReflectionClass($this->callable[0]))->getMethod($this->callable[1]);
+
+            } else {
+                $reflFunc = new ReflectionFunction($this->callable);
+
+            }
+
+            $args = array_map(fn(\ReflectionParameter $param) => $param->getName(), $reflFunc->getParameters());
+
+            $argsValue = array_map(
+                function (string $name) use ($parameters) {
+                    return $parameters[$name];
+                },
+                $args
+            );
```
Utiliser ReflectionMethod pour vérifier si la méthode a besoin de la request. On part du principe que les variables dans une variable ont le même nom que les arguments d'une action d'un controleur. Donc tableau associatif simple

TBoileau commented on src/Router/Route.php:
```diff
@@ -4,56 +4,111 @@
-    }
+        $pattern = str_replace('/', '\/', $this->path);
+        $pattern = sprintf('/^%s$/', $pattern);
+        $pattern = preg_replace('/({\w+})/', '(.+)', $pattern);
```
[Cette étape à décomposer avec la notion de requirements](https://github.com/TBoileau/php-poo-mvc/blob/master/app/Router/Route.php#L85)

TBoileau commented on src/Router/Route.php:
```diff
@@ -4,56 +4,111 @@
+use ReflectionException;
+use ReflectionFunction;
+
+class Route
```
Remettre le final

TBoileau commented on src/Router/Router.php:
```diff
@@ -4,53 +4,92 @@
+use Blog\Router\Exceptions\RouteNotFoundException;
 
-final class Router
+class Router
```
Remettre final

TBoileau replied:
Ajouter les 2 iinterfaces :
RouterInterface
UrlGeneratorInterface

TBoileau commented on src/Router/Router.php:
```diff
@@ -4,53 +4,92 @@
+     * @throws RouteNotFoundException
+     * @throws \ReflectionException
+     */
+    public function call(string $path)
```
Passer plutôt un objet Request

TBoileau commented on src/Router/Route.php:
```diff
@@ -4,56 +4,111 @@
-    public function getParameters(): array
-    {
-        return $this->parameters;
+        return (bool)preg_match($pattern, $path);
```
Ne pas avoir à répéter le même traitement dans "test" et "call"

TBoileau commented on src/Router/Route.php:
```diff
@@ -4,56 +4,111 @@
+     * @return false|mixed
+     * @throws ReflectionException
+     */
+    public function call(string $path)
```
Ne pas passer de string mais une Request

TBoileau commented on tests/RouterTest.php:
```diff
@@ -0,0 +1,100 @@
+
+namespace Blog\Tests;
+
+use Blog\Fixtures\FooController;
+use Blog\Fixtures\HomeController;
```
Déplacer tes fixtures dans le dossier Tests

aespasa committed:
Implementing router v1.1

aespasa committed:
Implementing router

Now accept requirements as args (e.g: /users/{id:\d+}).

Request and Response are send automatically if the controller (callable) use them.

Configuration file is now a php one instead of a json.

Rewrite tests (not finished yet).

urlGenerator is not implemented yet.

TBoileau commented on composer.json:
```diff
@@ -9,7 +9,8 @@
     "psr-4": {
-      "Blog\\": "src/"
+      "Blog\\": "src/",
+      "Tests\\":  "tests/"
```
a mettre dans "autoload-dev"

TBoileau commented on tests/Fixtures/FooController.php:
```diff
@@ -0,0 +1,69 @@
+    {
+        $response->setContent("Hello $firstname $lastname!");
+
+        return $response->send();
```
Ce n'est pas le contrôleur qui envoie la réponse, mais l'index.php

TBoileau commented on src/Router/Router.php:
```diff
@@ -84,12 +111,36 @@ public function match(string $path): Route
+     */
+    public function call(string $path): mixed
+    {
+        $response = new Response();
```
ce n'est pas au routeur d'avoir la responsabilité de créer la réponse mais au contrôleur

TBoileau commented on src/Router/UrlGenerator.php:
```diff
@@ -0,0 +1,48 @@
+final class UrlGenerator
+{
+    public function __construct(
+        private array $routes
+    ) {
+    }
+
+    public function generate(string $name, array $parameters = []): string
+    {
+//        if ($this->routes->offsetExists($name) === false) {
+//            throw new \InvalidArgumentException(
+//                sprintf('Unknown %s name route', $name)
+//            );
+//        }
+//        $route = $this->routes[$name];
+//        if ($route->hasVars() && $parameters === []) {
+//            throw new \InvalidArgumentException(
+//                sprintf('%s route need parameters: %s', $name, implode(',', $route->getVarsNames()))
+//            );
+//        }
+//
+//        return self::resolveUri($route, $parameters);
+    }
+
+    private static function resolveUri(Route $route, array $parameters): string
+    {
+//        $uri = $route->getPath();
+//        foreach ($route->getVarsNames() as $variable) {
+//            $varName = trim($variable, '{\}');
+//            if (array_key_exists($varName, $parameters) === false) {
+//                throw new \InvalidArgumentException(
+//                    sprintf('%s not found in parameters to generate url', $varName)
+//                );
+//            }
+//            $uri = str_replace($variable, $parameters[$varName], $uri);
+//        }
+//
+//        return $uri;
+    }
+}
```
A supprimer

TBoileau commented on src/Router/Router.php:
```diff
@@ -6,14 +6,32 @@
+ * Class Router
+ * @package Blog\Router
+ */
+final class Router implements RouterInterface
```
Ajout UrlGeneratorInterface

aespasa committed:
Implement the router and routes.

Implement UrlGeneratorInterface.

Implement all tests in RouterTest.php, it must be run after all modifications.

The index file is the entry point. It generates the router then loads the routes inside it. Next, the router->call() is called with the Request object. If a route matches, its controller is called and it returns a Response. After all, the response is sent from the index.

When loading a route, we can pass parameters and expected regex, e.g: '/users/{id:\d+}' will look only for digits as the id parameter. If the parameter is incorrect, it returns a RouteNotFoundException.

aespasa committed:
Refactor to implement I(nterface) S(egregation) P(rinciple)
